mscore/pianoroll/pianoroll.cpp

```diff
@@ -100,74 +100,166 @@ PianorollEditor::PianorollEditor(QWidget* parent)
             b->setPalette(p);
             QAction* a = getAction(voiceActions[i]);
             b->setDefaultAction(a);
-            tb->addWidget(b);
+            tbMain->addWidget(b);
             }
 
-      tb->addSeparator();
+      tbMain->addSeparator();
 
       partLabel = new QLabel("Part:");
-      tb->addWidget(partLabel);
+      tbMain->addWidget(partLabel);
 
-      tb = addToolBar("Toolbar 2");
+      // --------------------------------------------------
+      // empty area for spacing
+
+      struct LenIconData
+      {
+          QString _icon;
+          int _len;
+          bool _selected;
+      };
+      LenIconData _iconData[] = {
+            { QStringLiteral(":/data/icons/note-1.svg"), 1, true },
+            { QStringLiteral(":/data/icons/note-2.svg"), 2, false },
+            { QStringLiteral(":/data/icons/note-4.svg"), 4, false },
+            { QStringLiteral(":/data/icons/note-8.svg"), 8, false },
+            { QStringLiteral(":/data/icons/note-16.svg"), 16, false },
+            { QStringLiteral(":/data/icons/note-32.svg"), 32, false },
+            { QStringLiteral(":/data/icons/note-64.svg"), 64, false },
+            { QStringLiteral(":/data/icons/note-128.svg"), 128, false },
+            { QStringLiteral(":/data/icons/note-256.svg"), 256, false },
+            { QStringLiteral(":/data/icons/note-512.svg"), 512, false },
+            { QStringLiteral(":/data/icons/note-1024.svg"), 1024, false },
+            { "", -1, false },
+            };
+
+      QToolBar* tbNoteLen = addToolBar("Toolbar Note Length");
+      QButtonGroup* bngrpNoteLen = new QButtonGroup();
+
+      for (LenIconData* p = _iconData; p->_len != -1; ++p) {
+            QToolButton* bnLen = new QToolButton();
+            QIcon icon;
+            icon.addFile(p->_icon, QSize(), QIcon::Normal, QIcon::Off);
+            bnLen->setIcon(icon);
+            bnLen->setCheckable(true);
+            int length = p->_len;
+            connect(bnLen, &QToolButton::clicked, this, [=](){this->setEditNoteLength(length);});
+
+            if (p->_selected)
+                  bnLen->setChecked(true);
+            bngrpNoteLen->addButton(bnLen);
+            tbNoteLen->addWidget(bnLen);
+            }
+
+      QToolButton* bnDot = new QToolButton();
+      QIcon icon;
+      icon.addFile(QStringLiteral(":/data/icons/note-dot.svg"), QSize(), QIcon::Normal, QIcon::Off);
+      bnDot->setIcon(icon);
+      bnDot->setCheckable(true);
+      connect(bnDot, &QToolButton::clicked, this, &PianorollEditor::setEditNoteDot);
+      tbNoteLen->addWidget(bnDot);
```

Comment:
what about the rest button?

Reply:
I can't think of an way to handle rests in the PRE.  Currently I have the ability to erase notes and let the engine decide how to link rests together.

Reply:
On second thought, maybe there could be an extra bar-like symbol in the edit window  that indicated rests, but that would be a big addition and outside of the scope of  this PR.

Reply:
I very well may have misunderstood what this PR is about... Just for the Piano Roll Editor?